### PR TITLE
Fsamin/template

### PR DESCRIPTION
**Pull request by fsamin**
feat (api): templates as go-plugin
feat (api): manage templates api handlers
feat (api): load action as HCL script
feat (api): Import function handler
feat (cli): admin templates management
feat (cli): apply template rework

### CODE REVIEW

File: engine/api/objectstore/types.go [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package objectstore
+
+type Object interface {
```
Comment (sguiheux) [inline engine/api/objectstore/types.go]: Doc

File: engine/api/project.go [hunk 2]
```diff
@@ -84,24 +85,25 @@ func updateProject(w http.ResponseWriter, r *http.Request, db *sql.DB, c *contex
...
 		WriteError(w, r, sdk.ErrInvalidProjectName)
 		return
 	}
 
 	// Check Request
-	if key != projectArg.Key {
-		log.Warning("updateProject: bad Project key %s/%s \n", key, projectArg.Key)
+	if key != proj.Key {
+		log.Warning("updateProject: bad Project key %s/%s \n", key, proj.Key)
 		w.WriteHeader(http.StatusBadRequest)
```
Comment (sguiheux) [inline engine/api/project.go]: sdk.WrongRequest => WriteError

File: engine/api/project.go [hunk 3]
```diff
@@ -205,28 +207,28 @@ func getProject(w http.ResponseWriter, r *http.Request, db *sql.DB, c *context.C
 }
 
 func addProject(w http.ResponseWriter, r *http.Request, db *sql.DB, c *context.Context) {
-
 	// Get body
 	data, err := ioutil.ReadAll(r.Body)
 	if err != nil {
 		w.WriteHeader(http.StatusBadRequest)
```
Comment (sguiheux) [inline engine/api/project.go]: WriteError  WrongRequest

File: engine/api/project.go [hunk 4]
```diff
@@ -248,27 +250,34 @@ func addProject(w http.ResponseWriter, r *http.Request, db *sql.DB, c *context.C
...
+		sessionKey, err := auth.NewSession(router.authDriver, c.User)
 		if err != nil {
+			log.Critical("Instance> Error while creating new session: %s\n", err)
+			WriteError(w, r, err)
+			return
+		}
+		//Create from wizard
+		if err := project.CreateFromWizard(db, p, c.User, sessionKey); err != nil {
 			log.Warning("AddProject: Cannot create project: %s\n", err)
 			w.WriteHeader(http.StatusInternalServerError)
```
Comment (sguiheux) [inline engine/api/project.go]: WriteError

File: engine/api/templates.go [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+	if err != nil {
+		log.Warning("fileUploadAndGetTemplate> unable to get template info: %s", err)
+		return nil, nil, nil, deferFunc, sdk.NewError(sdk.ErrPluginInvalid, err)
+	}
+
+	return ap, params, content, deferFunc, nil
+}
+
+func getTemplatesHandler(w http.ResponseWriter, r *http.Request, db *sql.DB, c *context.Context) {
+	dbmap := database.DBMap(db)
```
Comment (sguiheux) [inline engine/api/templates.go]: go to api/template/template.go

File: engine/api/templates.go [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+	if err != nil {
+		log.Warning("applyTemplatesHandler> Cannot load project %s: %s\n", projectKey, err)
+		WriteError(w, r, err)
+		return
+	}
+
+	// Get data in body
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		WriteError(w, r, err)
```
Comment (sguiheux) [inline engine/api/templates.go]: sdk.ErrWrongRequest

File: engine/api/templates.go [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+		log.Warning("applyTemplateOnApplicationHandler> Cannot load application %s: %s\n", appName, err)
+		WriteError(w, r, err)
+		return
+	}
+
+	// Get data in body
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log.Warning("applyTemplateOnApplicationHandler> Unable to read body : %s\n", err)
+		WriteError(w, r, err)
```
Comment (sguiheux) [inline engine/api/templates.go]: WrongRequest?

File: engine/api/templates_test.go [hunk 8]
```diff
@@ -0,0 +1,841 @@
...
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"mime/multipart"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"testing"
+
```
Comment (sguiheux) [inline engine/api/templates_test.go]: lot of line return

File: engine/api/objectstore/openstack.go [hunk 9]
```diff
@@ -114,17 +57,18 @@ func (ops *OpenstackStore) Status() string {
...
 
-// DeleteArtifact removes artifact data from openstack
-func (ops *OpenstackStore) DeleteArtifact(art sdk.Artifact) error {
-	container, object := ops.format(art.Name, art.Project, art.Application, art.Environment, art.Pipeline, art.Tag)
-	log.Info("OpenstackStore> Deleting /%s/%s\n", container, object)
-
-	return deleteObject(ops.token.ID, ops.endpoint, container, object)
+// Delete should delete on openstack
+func (ops *OpenstackStore) Delete(o Object) error {
+	return nil
```
Comment (proullon) [inline engine/api/objectstore/openstack.go]: ?

Comment (fsamin) [inline engine/api/objectstore/openstack.go]: fixed : #68

**Pull request merged**